### PR TITLE
Product Categories List: Add show child categories only toggle

### LINKED ISSUE #6346: Product Categories: Show children of current category.
## Is your feature request related to a problem? Please describe.
The classic widget which this block has replaced had the option to display only the children of the current category. The Product Categories block does not have this functionality. Since this is supposed to be a replacement it makes sense to strive for parity between the two.

## Describe the solution you'd like
Add an option to the Product Categories block so that when it is used as a widget, merchants can select an option to show only children of the current category the user is on.

| Classic Widget | Current Block |
| --- | ----------- |
| image (2) | Screenshot 2022-04-29 at 11 06 06 |

### PR DESCRIPTION
**Pull request by tjcafferkey**
Add a "Only show children of current category" toggle for the Product Categories List block. This will only work on product category pages (e.g. when used on the Product Category template).

Fixes https://github.com/woocommerce/woocommerce-blocks/issues/6346

#### Automated Tests
* [ ] Changes in this PR are covered by Automated Tests.
  * [ ] Unit tests
  * [ ] E2E tests

#### User Facing Testing

1. Set up product categories with children and grandchildren, making sure they have products.
2. Add the Product Categories List block to a page, and also to the Products by Category template.
3. Check that the toggle only appears when the block is used in the Site Editor (e.g. on a template)
4. Ensure that when the toggle is enabled and you're visiting a category with children, only the children are shown.
5. Make sure that if when you have the _show children only_ toggle enabled, and you click the last child then it renders the current category instead of rendering nothing at all.
6. Ensure no regressions when using this block on posts/pages
7. Enable Classic theme such as Storefront
8. Usage the block in the widget area and check that the feature is available and works as expected and described above.

* [ ] Do not include in the Testing Notes 

### WooCommerce Visibility

* [x] WooCommerce Core
* [ ] Feature plugin
* [ ] Experimental

### Changelog

> Product Categories List: Add "Show child categories only" toggle.

### CODE REVIEW

File: assets/js/blocks/product-categories/block.js [hunk 1]
```diff
@@ -142,6 +152,24 @@ const ProductCategoriesBlock = ( { attributes, setAttributes, name } ) => {
...
 						}
 					/>
+					{ isSiteEditor && (
+						<ToggleControl
+							label={ __(
+								'Only show children of current category',
+								'woo-gutenberg-products-block'
+							) }
+							help={ __(
+								'This will effect product category pages',
```
Comment (Aljullu) [inline assets/js/blocks/product-categories/block.js]: I think there is a typo here:

```suggestion
								'This will affect product category pages',
```

File: src/BlockTypes/ProductCategories.php [hunk 2]
```diff
@@ -77,7 +79,7 @@ protected function render( $attributes, $content, $block ) {
 			if ( strstr( $content, 'data-is-dropdown="true"' ) ) {
 				$attributes['isDropdown'] = true;
 			}
-			if ( strstr( $content, 'data-is-hierarchical="false"' ) ) {
+			if ( strstr( $content, 'data-i\s-hierarchical="false"' ) ) {
```
Comment (Aljullu) [inline src/BlockTypes/ProductCategories.php]: Is this change intentional?

File: assets/js/blocks/product-categories/block.js [hunk 3]
```diff
@@ -142,6 +152,24 @@ const ProductCategoriesBlock = ( { attributes, setAttributes, name } ) => {
 							setAttributes( { hasEmpty: ! hasEmpty } )
 						}
 					/>
+					{ isSiteEditor && (
```
Comment (Aljullu) [inline assets/js/blocks/product-categories/block.js]: This option will be available when adding the block to any template in the site editor, even if it's not the Product Categories one (ie: when adding the block to the _Home_ template). I don't know if this is an issue, but I wanted to mention just in case.

Another thing to consider is whether we should enable this option when adding the block to the _Widgets_ screen in classic themes.

File: src/BlockTypes/ProductCategories.php [hunk 4]
```diff
@@ -134,15 +136,37 @@ protected function get_container_classes( $attributes = array() ) {
...
+				]
+			);
+		}
+
+		// If they have $children_only enabled, and the current category is the last child in the hierarchy
+		// we need to show the current category instead of rendering nothing at all.
+		if ( $children_only && ( ! is_array( $categories ) || empty( $categories ) ) ) {
+			$current_category = get_term( $term_id, 'product_cat' );
+			return [ $current_category ];
+		}
```
Comment (Aljullu) [inline src/BlockTypes/ProductCategories.php]: Have alternatives been considered? I'm not saying I don't like this approach, but I think I would have expected it to either display nothing or display the current category + all siblings.

**Review by Aljullu — COMMENTED** [review]
This is working as described and code changes look good. I just left some inline questions about how some things work.

File: src/BlockTypes/ProductCategories.php (hunk 2 above)
Comment (tjcafferkey) [reply src/BlockTypes/ProductCategories.php]: Hmm no, dunno how that happened must have been during some auto-linting fixes maybe? 🤔

File: assets/js/blocks/product-categories/block.js (hunk 3 above)
Comment (tjcafferkey) [reply assets/js/blocks/product-categories/block.js]: >I don't know if this is an issue, but I wanted to mention just in case.

Yeah, this is a known issue which is why I put the tooltip in the `help` prop on the editor side. I didn't want to overly complicate this since I don't think there's a standard way to approach this. We'd also need to consider all kinds of category templates e.g. category specific ones.

>Another thing to consider is whether we should enable this option when adding the block to the Widgets screen in classic themes.

Yes, good point! I will address this in this PR.

File: src/BlockTypes/ProductCategories.php (hunk 4 above)
Comment (tjcafferkey) [reply src/BlockTypes/ProductCategories.php]: Personally, I did consider rendering nothing but on templates where the category block is the only block is a sidebar style template it may look weird rendering an empty sidebar which is probably my main reason for it. Maybe we should ask Shaun for input here:

@shaunandrews in summary:

I am introducing a feature to the Product Categories block where the user can toggle an option to only render the child categories of the current category that is being browsed. This is for feature parity with the classic widget version which you can see a little bit more on here https://github.com/woocommerce/woocommerce-blocks/issues/6346

In the scenario where you on a category with no children (e.g. the last category in the hierarchy) do we want to render the current category they're on, or render nothing at all?

File: assets/js/blocks/product-categories/block.js (hunk 3 above)
Comment (tjcafferkey) [reply assets/js/blocks/product-categories/block.js]: Done in https://github.com/woocommerce/woocommerce-blocks/pull/9550/commits/73008c0a138ea3cf1f7f3d1910f2d7833b1e1488

File: src/BlockTypes/ProductCategories.php (hunk 4 above)
Comment (shaunandrews) [reply src/BlockTypes/ProductCategories.php]: Matching the behavior of the existing widget is a solid place to start. That said, I'm not entirely sure I understand the desired outcome, nor the intended use-case.

Any chance you can include some visual for how the block will behave/render in each scenario? For instance, does the current category always show, or will it only show if no other categories exist? It seems like it might be strange to only show the current category with some relatively obscure logic.

Comment (tjcafferkey) [reply src/BlockTypes/ProductCategories.php]: >Matching the behavior of the existing widget is a solid place to start.

Yeah, I agree. I've just reactivated the classic widget locally and done a side-by-side test which I've screen recorded below. In the below video I'm on the "Accessories" product category which has a "Grandchild" sub-category I created which is last in the hierachy, so this is the only one that shows initially. Going to this category results in it disappearing because there are no more sub-categories. Based on this I'll alter the block features logic to mirror this and if there are any further improvements we want to make we can do so in a follow up PR.

https://github.com/woocommerce/woocommerce-blocks/assets/8639742/d7f09b3b-b817-455f-bf2d-f66300e954f5

Comment (tjcafferkey) [reply src/BlockTypes/ProductCategories.php]: @Aljullu I've remove the logic in question here https://github.com/woocommerce/woocommerce-blocks/pull/9550/commits/00feab1d63b5e0fe2692abab937b12d3c5746078

Something worth mentioning is that if you add a `Title` block before the `Product Categories List` block then when it renders nothing, the title block will still appear.

We fixed this in other blocks such as the filter blocks but this is because those blocks originally _came with_ a title when inserted, whereas the `Product Categories List` block does not. So there is no relationship between the two blocks.

**Review by Aljullu — APPROVED** [review]
Looks good on my end, thanks for the updates!
